core/src/main/scala/org/apache/spark/deploy/master/Master.scala

```diff
@@ -924,49 +939,74 @@ private[deploy] class Master(
    * Return the UI if successful, else None
    */
   private[master] def rebuildSparkUI(app: ApplicationInfo): Option[SparkUI] = {
+    val futureUI = asyncRebuildSparkUI(app)
+
+    // Wait until the UI is rebuilt or cancelled, log status occasionally for apps with many events
+    var waitSec = 2
+    do {
+      try {
+        val ui = Await.result(futureUI, Duration(waitSec, TimeUnit.SECONDS))
```

Comment:
won't you end up calling `Await.result` every time we throw timeout exception? I think we only want to await the future once